src/CodeGen_OpenCL_Dev.cpp

```diff
@@ -313,6 +461,35 @@ void CodeGen_OpenCL_Dev::CodeGen_OpenCL_C::visit(const Cast *op) {
     }
 }
 
+void CodeGen_OpenCL_Dev::CodeGen_OpenCL_C::visit(const Select *op) {
+    if (op->condition.type().is_vector()) {
+        // We eliminated boolean vectors in OpenCL C. In the process
+        // of doing so, we replaced the vector conditions of selects
+        // with integer vectors, compared with 0. To make legal OpenCL
+        // C, we just need to grab the integer vector out of the
+        // comparison with zero.
+        Expr condition = op->condition;
+        const NE* cond_neq = condition.as<NE>();
+        if (cond_neq) {
+            if (is_zero(cond_neq->b)) {
+                condition = cond_neq->a;
+            }
+        }
```

Comment:
I'd like to keep it optional. The reason is maybe a bit hacky: currently, CodeGen_C implements many intrinsic functions by generating IR on the fly. This IR doesn't get processed by EliminateBoolVectors. As long as these expressions generated by CodeGen_C don't involve a logical op that isn't itself the select condition, just optionally stripping away the != 0 handles these cases.

However, I did add an assert after this to ensure the condition is in fact an integer vector. This will catch it if we ever do add an implementation of an intrinsic to CodeGen_C that requires bool vector elimination beyond the "root" of the condition.